src/node/desktop/src/main/gwt-callback.ts

```diff
@@ -91,19 +91,46 @@ export class GwtCallback extends EventEmitter {
       }
     });
 
-    ipcMain.handle('desktop_get_save_file_name', (event, caption: string, label: string,
+    ipcMain.handle('desktop_get_save_file_name', async (event, caption: string, label: string,
       dir: string, defaultExtension: string, forceDefaultExtension: boolean,
```

Comment:
You're not using `defaultExtension` or `forceDefaultExtension`; those are used in the current implementations and need to be reimplemented if possible on Electron. Probably involving `filters: FileFilter[]`, but I didn't look very closely. Be sure to checkout how it's done in both the Linux/Windows implementation (DesktopGwtCallback.cpp) and Mac (DesktopGwtCallbackMac.mm). The Qt docs are at https://doc.qt.io/qt-5/qfiledialog.html

Some of the resulting behavior may be easier to see/troubleshoot on Windows.

Reply:
I couldn't see a clear way to implement the default extension, as the `filters:` option requires a `name:` in addition to the extension. The example from electron's documentation gives something like:

```
{
  filters: [
    { name: 'Images', extensions: ['jpg', 'png', 'gif'] },
    { name: 'Movies', extensions: ['mkv', 'avi', 'mp4'] },
    { name: 'Custom File Type', extensions: ['as'] },
    { name: 'All Files', extensions: ['*'] }
  ]
}
```

Since a `name` for the extension isn't included in this call from GWT, the best I could think of was to create some static filter object like the above that holds all filetypes/extensions we work with and sort them into categories, but that sounds tedious to create and even worse to maintain.

All of that seemed moot to me anyway, because when you have a document open, and trigger this "Save As..." dialog, the extension is already present, and left there by default (the filename, not including the file extension, is automatically highlighted by default, ready to be changed by typing), which is what I'm assuming the filter was supposed to do in Qt anyway:
![Screen Shot 2021-08-05 at 10 03 56 AM](https://user-images.githubusercontent.com/56326543/128383042-2fdaefe9-f5d8-4084-a3d7-df6ce1b07d34.png)

This seemed reasonable to me, so I didn't spend any more time on it. With this extra info, is leaving it as-is okay with you?

Reply:
Sorry, `filters` was a red herring (I wasn't looking at the code), as I do use it for `desktop_get_open_file_name`.

We definitely need the behavior of the current code to be ported over. Here's an example why:

* File / New File / R Script
* Save
* Type "hello" and hit Enter

In RStudio (shipped version) this will save as "hello.R". In the Electron version (without `defaultExtension` implementation), the file tries to save as plain `hello`, prompting a warning message.

This behavior is implemented AFTER the dialog is dismissed (again, sorry to confuse matters with mention of `filters`), here: https://github.com/rstudio/rstudio/blob/5dd6fdfb5c912e7afc0975c78dd4f313722bd5dd/src/cpp/desktop/DesktopGwtCallback.cpp#L298-L323

Reply:
Ah! I see now. 

I guess `filters` wasn't entirely a red herring, because you need `filters` to get the `.R` extension to show up for a new, untitled document